crates/composefs/src/tree.rs

```diff
@@ -1,471 +1,32 @@
-    ///
-    ///  * `filename`: the filename in the current directory.  If you need to support full
-    ///    pathnames then you should call `Directory::split()` first.
-    pub fn remove(&mut self, filename: &OsStr) {
-        self.entries.remove(filename);
-    }
-
-    /// Does a directory lookup on the given filename, returning the Inode if it exists.
-    ///
-    /// # Arguments
-    ///
-    ///  * `filename`: the filename in the current directory.  If you need to support full
-    ///    pathnames then you should call `Directory::split()` first.
-    pub fn lookup(&self, filename: &OsStr) -> Option<&Inode<ObjectID>> {
-        self.entries.get(filename)
-    }
-
-    /// Removes an item from the directory, if it exists, returning the Inode value.
-    ///
-    /// # Arguments
-    ///
-    ///  * `filename`: the filename in the current directory.  If you need to support full
-    ///    pathnames then you should call `Directory::split_mut()` first.
-    pub fn pop(&mut self, filename: &OsStr) -> Option<Inode<ObjectID>> {
-        self.entries.remove(filename)
-    }
-
-    /// Removes all content from this directory, making the directory empty.  The `stat` data
-    /// remains unmodified.
-    pub fn clear(&mut self) {
-        self.entries.clear();
-    }
-
-    pub fn newest_file(&self) -> i64 {
-        let mut newest = self.stat.st_mtim_sec;
-        for inode in self.entries.values() {
-            let mtime = match inode {
-                Inode::Leaf(ref leaf) => leaf.stat.st_mtim_sec,
-                Inode::Directory(ref dir) => dir.newest_file(),
-            };
-            if mtime > newest {
-                newest = mtime;
-            }
-        }
-        newest
-    }
-}
-
-#[derive(Debug)]
-pub struct FileSystem<ObjectID: FsVerityHashValue> {
-    pub root: Directory<ObjectID>,
-    pub have_root_stat: bool,
-}
-
-impl<ObjectID: FsVerityHashValue> Default for FileSystem<ObjectID> {
-    fn default() -> Self {
-        Self {
-            root: Directory::default(),
-            have_root_stat: false,
-        }
-    }
-}
-
-impl<ObjectID: FsVerityHashValue> FileSystem<ObjectID> {
-    pub fn set_root_stat(&mut self, stat: Stat) {
-        self.have_root_stat = true;
-        self.root.stat = stat;
-    }
-
-    pub fn ensure_root_stat(&mut self) {
-        if !self.have_root_stat {
-            self.root.stat.st_mtim_sec = self.root.newest_file();
-            self.have_root_stat = true;
-        }
-    }
-}
+// Re-export generic types. Note that we don't need to re-write
+// the generic constraint T: FsVerityHashValue here because it will
+// be transitively enforced.
+pub type LeafContent<T> = generic_tree::LeafContent<RegularFile<T>>;
+pub type Leaf<T> = generic_tree::Leaf<RegularFile<T>>;
+pub type Directory<T> = generic_tree::Directory<RegularFile<T>>;
+pub type Inode<T> = generic_tree::Inode<RegularFile<T>>;
+pub type FileSystem<T> = generic_tree::FileSystem<RegularFile<T>>;
 
 #[cfg(test)]
 mod tests {
```

Comment:
In theory we actually don't *really* need any tests at all here I think; we get coverage from the generic tests and higher level tests. But I kept a few just to prove things out.